File: scenarios/haskell/hs-binary-parser/template/src/BinaryParser.hs
module BinaryParser (parsePacket, Packet(..)) where

import qualified Data.ByteString as B
import Data.Int
import Data.Word

data Packet = Login { username :: String, hash :: B.ByteString }
            | Order { packetId :: Int64, symbol :: String, side :: Word8, price :: Double, qty :: Word32 }
            | Heartbeat { timestamp :: Int64 }
            deriving (Show, Eq)

parsePacket :: B.ByteString -> Either String Packet
parsePacket _ = Left "Not implemented"

File: scenarios/ocaml/ml-binary-parser/template/src/binary_parser.ml
type packet = 
  | Login of { username: string; hash: bytes }
  | Order of { id: int64; symbol: string; side: int; price: float; qty: int32 }
  | Heartbeat of { timestamp: int64 }

let parse_packet (data: bytes) : (packet, string) result =
  reqtype Error "Not implemented"
